### PR TITLE
Reference templates

### PR DESCRIPTION
**Pull request by ronpal**
# Description

The built-in package "reference" gives the user a module that shows properly formatted yaml files for all supported resource types.

The package is available through modules init

## Checklist

- [ ] Tests added/updated.
- [ ] Run Demo Job Locally.
- [ ] Documentation updated.
- [ ] Changelogs updated in [CHANGELOG.cdf-tk.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.cdf-tk.md).
- [ ] Template changelogs updated in [CHANGELOG.templates.md](https://github.com/cognitedata/toolkit/blob/main/CHANGELOG.templates.md).
- [ ] Version bumped.
  [_version.py](https://github.com/cognitedata/toolkit/blob/main/cognite/cognite_toolkit/_version.py) and
  [pyproject.toml](https://github.com/cognitedata/toolkit/blob/main/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 1]
```diff
@@ -0,0 +1,338 @@
+import inspect
+import itertools
+import re
+import shutil
+
+# from cognite.client.data_classes.transformations.common import TransformationDestination
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:6]:
```suggestion
```

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+from cognite.client.data_classes.iam import Capability
+from cognite.client.data_classes.transformations import OidcCredentials
+from cognite.client.data_classes.transformations.common import TransformationDestination
+from cognite.client.utils._text import to_camel_case
+
+from cognite_toolkit._cdf_tk.loaders import ResourceLoader
+from cognite_toolkit._cdf_tk.validation import validate_resource_yaml
+
+
+# * in a line signifies that the line is a comment, for example where the config has multiple options
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:23]:
```suggestion
# * in a line signifies that the line is a comment, for example, where the config has multiple options
```

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+    "dict[str, str]",
+    "list[str]",
+    "Any",
+    "Filter",
+    "list[ExternalId]",
+    "'allUserAccounts'",
+    "list[ViewId]",
+    "dict[str, ViewPropertyApply]",
+    # "list[ViewId | ViewApply]",
+]
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:52]: O(1) better than O(n) Lookup
```suggestion
IGNORED_ANNOTATIONS = {
    "str",
    # "int",
    "float",
    "bool",
    "dict",
    "list",
    "None",
    "enum",
    "dict[str, str]",
    "list[str]",
    "Any",
    "Filter",
    "list[ExternalId]",
    "'allUserAccounts'",
    "list[ViewId]",
    "dict[str, ViewPropertyApply]",
}
```

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 4]
```diff
@@ -0,0 +1,338 @@
+import inspect
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:1]: Need to support Python 3.9
```suggestion
from __future__ import annotations

import inspect
```

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+                continue
+            for param in get_parameters(cls):
+                resolved[to_camel_case(param.name)] = unpack_annotations(str(param.annotation))
+
+        return resolved
+
+    return str_annotations
+
+
+def expand_parameters(loader_cls: Any, cls: Any, optional: bool = False) -> dict[str, Any]:
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:139]: This feels like reinventing the wheel. I think the same can be achieved with the parameter module with some modifications. But less discuss whether it is worth the effort.

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+        # special case for groups: want caps at the end
+        if "capabilities" in ref:
+            caps_yaml = comment_optional(yaml.dump({"capabilities": ref.pop("capabilities")}, sort_keys=False))
+            main_yaml = comment_optional(yaml.dump(ref, sort_keys=False))
+            final_yaml = f"{main_yaml}{caps_yaml}"
+
+        folder = target_path / loader.folder_name
+        file_name = folder / Path(f'reference.{write_cls.__name__.replace("Write","").replace("Apply","")}.yaml')
+        Path.mkdir(folder, exist_ok=True)
+        Path.write_text(file_name, final_yaml)
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:292]: `loader_cls` to make it clear that this is a a class, and we already have `kind` for this exact purpose
```suggestion
    for loader_cls in [
        loader for loader in ResourceLoader.__subclasses__() if loader.__name__ not in ["ResourceContainerLoader"]
    ]:
        write_cls = loader_cls.resource_write_cls  # type: ignore

        ref = expand_parameters(loader_cls, write_cls)

        final_yaml = comment_optional(yaml.dump(ref, sort_keys=False))

        # special case for groups: want caps at the end
        if "capabilities" in ref:
            caps_yaml = comment_optional(yaml.dump({"capabilities": ref.pop("capabilities")}, sort_keys=False))
            main_yaml = comment_optional(yaml.dump(ref, sort_keys=False))
            final_yaml = f"{main_yaml}{caps_yaml}"

        folder = target_path / loader_cls.folder_name
        file_name = folder / Path(f'reference.{loader_cls.kind}.yaml')
        Path.mkdir(folder, exist_ok=True)
        Path.write_text(file_name, final_yaml)

        print(f"Wrote {file_name}")
```

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py [hunk 7]
```diff
@@ -0,0 +1,338 @@
...
+        )
+
+        total_warnings += len(warnings)
+        if len(warnings) > 0:
+            print(f"Warnings for {file_name}")
+            for warning in warnings:
+                print(warning)
+            break
+
+    print(f"Total warnings: {total_warnings}")
```
Comment (doctrino) [inline cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:318]:
```suggestion
    total_warnings = 0
    for loader_cls in [
        loader for loader in ResourceLoader.__subclasses__() if loader.__name__ not in ["ResourceContainerLoader"]
    ]:
        folder = target_path / loader_cls.folder_name
        file_name = folder / Path(f'reference.{loader_cls.kind}.yaml')
        warnings = validate_resource_yaml(
            data=yaml.safe_load(Path.read_text(file_name)),
            spec=loader_cls.get_write_cls_parameter_spec(),
            source_file=file_name,
        )

        total_warnings += len(warnings)
        if len(warnings) > 0:
            print(f"Warnings for {file_name}")
            for warning in warnings:
                print(warning)
            break

    print(f"Total warnings: {total_warnings}")
```

**Review by doctrino — COMMENTED** [review]
First impression is that this is well worth an experiment. I see how this is different than example resources, and is useful to generate.

I am not to happy with `expand_parameters` but lets discuss that and the reason for why you wrote it instead of using `loader_cls.get_write_cls_parameter_spec()`.

File: cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py (hunk 5 above)
Comment (ronpal) [reply cognite_toolkit/_cdf_tk/prototypes/_packages/update_references.py:139]: As discussed IRL, you're right. This is a temporary solution.

**Review by doctrino — APPROVED** [review]
I think this code can be coordinated with the _parameters subpackages. But that is an effort for another day.

**Pull request merged**